nemo_text_processing/g2p/data/heteronym_classification_data.py

```diff
@@ -65,98 +67,136 @@ def __init__(
+                        "subtokens_mask": example[1],
+                        "target": example[2],  # None if self.with_labels is False
                     }
-                )
-            else:
-                target_and_negatives, subword_mask = self._prepare_sample(sentence, start, end, homograph)
-                self.data.append(
-                    {"input": sentence, "subword_mask": subword_mask, "target_and_negatives": target_and_negatives}
-                )
+                    self.data.append(example_dict)
 
         logging.info(f"Number of samples in {manifest}: {len(self.data)}, remove {num_skipped} lines")
 
-    def _prepare_sample(self, sentence: str, start: int, end: int, homograph: str, word_id=None):
+    def _prepare_sample(
+        self,
+        sentence: str,
+        start_end: List[Tuple[int, int]],
+        homographs: List[str],
+        word_ids: Optional[List[int]] = None,
+    ):
         """
         Prepares a single training sample
 
         Args:
             sentence: input sentence in grapheme form
-            start: start of the homograph span
-            end: end of the homograph span
-            homograph: homograph
-            word_id: [Optional] target word_id, use None for inference
+            start_end: start and end indices of the homograph spans, start_end indices should be in increasing order
+            homographs: homographs present in the sentence
+            word_ids: [Optional] target word_ids, use None for inference
         """
-        l_context = self.tokenizer.text_to_ids(sentence[:start])
-        r_context = self.tokenizer.text_to_ids(sentence[end:])
-        l_context_len = len(l_context)
-        r_context_len = len(r_context)
-        sentence_tokenized = self.tokenizer.text_to_ids(sentence)
-
-        grapheme_ipa_forms = self.wiki_homograph_dict[homograph.lower()]
-        homograph_len = len(sentence_tokenized[l_context_len : len(sentence_tokenized) - r_context_len])
-        # use prediction of the first subword of the homograph
-        homograph_mask = [1] + [0] * (homograph_len - 1)
-        subword_mask = [0] * (l_context_len + 1) + homograph_mask + [0] * (r_context_len + 1)
-        target_and_negatives = [self.wordid_to_idx[wordid_] for wordid_ in grapheme_ipa_forms]
-        output = [target_and_negatives, subword_mask]
+        # drop example where sequence length exceeds max sequence length, +2 for special tokens
+        length = len(self.tokenizer.text_to_tokens(sentence)) + 2
+        if length > self.max_seq_len:
+            logging.debug(f"Sequence length exceeds max sequence length ({self.max_seq_len}): {sentence}.")
+            return None
+
+        # check the correctness on start-end indices
+        for homograph_, start_end_ in zip(homographs, start_end):
+            if homograph_.lower() != sentence[start_end_[0] : start_end_[1]].lower():
+                return None
+
+        # add bos token
+        input_ids = [self.tokenizer.bos_id]
```

Comment:
nit: `SOS` and `EOS` are general terms for start of sentence and end of sentence, respectively.